design/one-pager-resource-connectivity-mvp.md

```diff
@@ -0,0 +1,656 @@
+kind: GKEInstance
+metadata:
+  namespace: crossplane-system
+  name: example
+spec:
+  nameFormat: mycoolname
+  description: My cool cluster
+  masterAuth:
+    username: cooluser
+    password: secretpassword
+    clientCertificateConfig:
+      issueClientCertificate: true
+  loggingService: logging.googleapis.com/kubernetes
+  monitoringService: monitoring.googleapis.com
+  network: mycoolnetwork
+  clusterIpv4Cidr: 10.0.0.0/8
+  addonsConfig:
+    httpLoadBalancing:
+      disabled: false
+    horizontalPodAutoscaling:
+      disabled: false
+    networkPolicyConfig:
+      disabled: false
+  subnetwork: mycoolsubnetwork
+  nodePools:
+  - name: string
+    config:
+      machineType: n1-standard-2
+      diskSizeGb: 20
+      oauthScopes:
+      - cool
+      serviceAccount: mycoolserviceaccount@accounts.google.com
+      metadata:
+        cool: very
+      imageType: coreos/stable
+      labels:
+        cool: true
+      localSsdCount: 0
+      tags:
+      - cool
+      preemptible: false
+      accelerators:
+        acceleratorCount: 2
+        acceleratorType: large
+      diskType: pd-ssd
+      minCpuPlatform: Intel Haswell
+      taints:
+        - key: color
+          value: purple
+          effect: NO_SCHEDULE
+    initialNodeCount: number
+    version: 1.14
+    autoscaling:
+      enabled: true
+      minNodeCount: 3
+      maxNodeCount: 9
+    management: object (NodeManagement)
+      autoUpgrade: true,
+      autoRepair: true,
+    maxPodsConstraint:
+      maxPodsPerNode: 254
+  locations:
+  - string
+  enableKubernetesAlpha: boolean
+  resourceLabels:
+    string: string
+  labelFingerprint: string
+  legacyAbac:
+    enabled: false
+  networkPolicy:
+    provider: CALICO
+    enabled: true
+  ipAllocationPolicy:
+    useIpAliases: true
+    createSubnetwork: false
+    subnetworkName: mycoolsubnetwork
+    clusterSecondaryRangeName: mycoolpodsrange
+    servicesSecondaryRangeName: mycoolservicesrange
+    clusterIpv4CidrBlock: 10.0.0.0/8
+    nodeIpv4CidrBlock: 172.16.1.0/24
+    servicesIpv4CidrBlock: 192.168.0.0/24
+    tpuIpv4CidrBlock: 172.16.2.0/24
+  masterAuthorizedNetworksConfig:
+    enabled: true
+    cidrBlocks:
+      displayName: coolcidr
+      cidrBlock: 93.80.0.0/16
+  maintenancePolicy: object (MaintenancePolicy)
+    window:
+      dailyMaintenanceWindow:
+        startTime: 01:30
+  defaultMaxPodsConstraint:
+    maxPodsPerNode: 254
+  resourceUsageExportConfig:
+    bigqueryDestination:
+      datasetId: mycooldataset
+    enableNetworkEgressMetering: true
+    consumptionMeteringConfig:
+      enabled: true
+  privateClusterConfig:
+    enablePrivateNodes: true
+    enablePrivateEndpoint: true
+    masterIpv4CidrBlock: 172.16.0.0/16
+  initialClusterVersion: 1.13
+  enableTpu: false
+```
+
+## Proposal
+
+Minimum viable support for resource connectivity can be enabled by ensuring
+three things.
+
+* All external resources required to configure connectivity have corresponding
```

Comment:
This is probably a dumb question, but which ones are missing today?

Reply:
Not a dumb question - I think we need more investigation to arrive at a definitive list. The idea is that this document will lead us to the set of connectivity external resources required to plumb a MySQL database to a Kubernetes cluster in each cloud provider, but there will be more as we plumb more external resource types together.

Reply:
I took a quick pass recently, trying to determine what external resources we might need to model for our current controllers. It's something like:

AWS
* Security groups
* Subnet groups
* DB subnet groups

Azure
* Resource group (done)
* Firewall rules
* Subnets

GCP
* Network
* Subnetwork
* Firewall rules

There's also AWS IAM roles, AWS EC2 keypairs, Azure service principals, and GCP service accounts, but they're all more credential-ey than connectivit-ey.

Reply:
I think it worth to point out that all of those address consumption needs **withing** the same cloud provider boundaries. I.e. both the compute resource and managed resource are hosted by the same provider.

Reply:
@ichekrygin Correct. Addressing inter-cloud connectivity is out of scope for the time being.